src/datadog/remote_config/capability.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include <cstdint>
+#include <string>
+
+#include "string_util.h"
+#include "string_view.h"
+
+namespace datadog {
+namespace remote_config {
+
+// Type alias for capabilities flags.
+//
+// Usage:
+//
+// using namespace datadog::remote_config::capability;
+// Capabilities c = APM_TRACING_SAMPLE_RATE | APM_TRACING_ENABLED;
+using Capabilities = uint64_t;
+
+// Type alias for product flags.
+//
+// Usage:
+//
+// using namespace datadog::remote_config::product;
+// Products p = AGENT_CONFIG | APM_TRACING;
+using Products = uint64_t;
+
+namespace capability {
+
+enum Flag : Capabilities {
+  // DEFAULT is a special value. It is the default of the product capability.
+  DEFAULT = 0,
+  ASM_ACTIVATION = 1 << 1,
+  ASM_IP_BLOCKING = 1 << 2,
+  ASM_DD_RULES = 1 << 3,
+  ASM_EXCLUSIONS = 1 << 4,
+  ASM_REQUEST_BLOCKING = 1 << 5,
+  ASM_RESPONSE_BLOCKING = 1 << 6,
+  ASM_USER_BLOCKING = 1 << 7,
+  ASM_CUSTOM_RULES = 1 << 8,
+  ASM_CUSTOM_BLOCKING_RESPONSE = 1 << 9,
+  ASM_TRUSTED_IPS = 1 << 10,
+  ASM_API_SECURITY_SAMPLE_RATE = 1 << 11,
+  APM_TRACING_SAMPLE_RATE = 1 << 12,
+  APM_TRACING_LOGS_INJECTION = 1 << 13,
+  APM_TRACING_HTTP_HEADER_TAGS = 1 << 14,
+  APM_TRACING_TAGS = 1 << 15,
+  ASM_PREPROCESSOR_OVERRIDES = 1 << 16,
+  ASM_CUSTOM_DATA_SCANNERS = 1 << 17,
+  ASM_EXCLUSION_DATA = 1 << 18,
+  APM_TRACING_ENABLED = 1 << 19,
+  APM_TRACING_DATA_STREAMS_ENABLED = 1 << 20,
+  ASM_RASP_SQLI = 1 << 21,
+  ASM_RASP_LFI = 1 << 22,
+  ASM_RASP_SSRF = 1 << 23,
+  ASM_RASP_SHI = 1 << 24,
+  ASM_RASP_XXE = 1 << 25,
+  ASM_RASP_RCE = 1 << 26,
+  ASM_RASP_NOSQLI = 1 << 27,
+  ASM_RASP_XSS = 1 << 28,
+  APM_TRACING_SAMPLE_RULES = 1 << 29,
+};
+
+}  // namespace capability
+
+#define DD_QUOTED_IMPL(ARG) #ARG
+#define DD_QUOTED(ARG) DD_QUOTED_IMPL(ARG)
+
+// List of remote configuration product built to support flag arithmetic
+#define DD_LIST_REMOTE_CONFIG_PRODUCTS \
+  X(AGENT_CONFIG, 1)                   \
+  X(AGENT_TASK, 2)                     \
+  X(APM_TRACING, 3)                    \
+  X(LIVE_DEBUGGING, 4)                 \
+  X(LIVE_DEBUGGING_SYMBOL_DB, 5)       \
+  X(ASM, 6)                            \
+  X(ASM_DD, 7)                         \
+  X(ASM_DATA, 8)                       \
+  X(ASM_FEATURES, 9)
+
+namespace product {
+
+enum Flag : Products {
```

Comment:
a flag is by definition a boolean, so the naming is semantically off. I get that they're assigned a specific bit, but that only makes an implementation of a set of products easier, it doesn't make them logically a "flag"

Reply:
Maybe, I am not fluent enough to argue. However, it is not uncommon to name bitwise enums `somethingFlags`. It is even a concept in C# ([doc](https://learn.microsoft.com/en-us/dotnet/api/system.flagsattribute?view=net-8.0), [example](https://code-corner.dev/2023/11/04/Understanding-Flag-Enums-in-C/))

Reply:
That is true enough, but the causation is the other way around, because it is a flag (a boolean), it tends to be a assigned a specific bit when making sets (bitsets) of these. If the product was used only in sets of products, to identify whether a certain product is present, then the usage would be acceptable. The [dictionary of computing](https://www.oxfordreference.com/display/10.1093/oi/authority.20110803095822224) defines a flag as:

> A variable whose value indicates the attainment of some designated state or condition by an item of equipment or a program. The flag is subsequently used as a basis for conditional branching and similar decision processes.

Say, in `extern _Atomic int interrupted; ...; while (!interrupted) { ... }`.

Anyway, at the end of the day it doesn't really matter, but I did find it confusing at first.